### PR TITLE
feat: detect when SMS are marked as spam or carrier is unavailable

### PR DESCRIPTION
**Pull request by AntoineAugusti**
Add log metric filters to detect when:
- SMS are marked as spam by a phone carrier
- when a phone carrier is unavailable

from SNS SMS log groups, in 2 AWS regions where we send SMS.

Add **warning alarms** using these new metrics:
- More than 10 SMS have been blocked as spam over 12 hours
- More than 10 SMS failed because a phone carrier is unavailable over 6 hours

Trello: https://trello.com/c/wzAeiXeO/365-be-alerted-when-phone-carriers-are-having-issues

### CODE REVIEW

**Comment by AntoineAugusti** [comment]
Had a chat with @willeybryan about thresholds and the fact that we don't want to be paged for that for now

File: aws/common/cloudwatch_log.tf [hunk 1]
```diff
@@ -46,6 +49,72 @@ resource "aws_cloudwatch_log_group" "sns_deliveries_failures_us_west_2" {
   }
 }
 
+###
+# AWS CloudWatch Logs Metrics
+###
+resource "aws_cloudwatch_log_metric_filter" "sns-sms-blocked-as-spam" {
+  name = "sns-sms-blocked-as-spam"
+  # See https://docs.amazonaws.cn/en_us/sns/latest/dg/sms_stats_cloudwatch.html#sms_stats_delivery_fail_reasons
+  pattern        = "{ $.delivery.providerResponse = \"Blocked as spam by phone carrier\" }"
```
Comment (AntoineAugusti) [inline aws/common/cloudwatch_log.tf:58]: See the syntax for JSON https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/FilterAndPatternSyntax.html

Tested with a sample payload.
Screen Shot 2021-04-11 at 15 36 58

File: aws/common/cloudwatch_alarms.tf [hunk 2]
```diff
@@ -144,6 +144,66 @@ resource "aws_cloudwatch_metric_alarm" "sns-sms-success-rate-canadian-numbers-us
...
+  metric_name         = aws_cloudwatch_log_metric_filter.sns-sms-blocked-as-spam.metric_transformation[0].name
+  namespace           = aws_cloudwatch_log_metric_filter.sns-sms-blocked-as-spam.metric_transformation[0].namespace
+  period              = 60 * 60 * 12
+  statistic           = "Sum"
+  threshold           = 10
+  alarm_actions       = [aws_sns_topic.notification-canada-ca-alert-warning.arn]
+  treat_missing_data  = "notBreaching"
+}
+
+resource "aws_cloudwatch_metric_alarm" "us-west-2-sns-sms-blocked-as-spam-warning" {
```
Comment (QuanNguyen-CDS) [inline aws/common/cloudwatch_alarms.tf:161]: historically the defacto naming convention is <warning description>-<[optional]:region>
e.g. `sns-sms-success-rate-canadian-numbers-us-west-2-critical`
The naming has been changed to <[optional]:region>-<warning description>
Is there a reason why we are reversing this?

File: aws/common/cloudwatch_alarms.tf [hunk 3]
```diff
@@ -144,6 +144,66 @@ resource "aws_cloudwatch_metric_alarm" "sns-sms-success-rate-canadian-numbers-us
...
+  metric_name         = aws_cloudwatch_log_metric_filter.sns-sms-phone-carrier-unavailable.metric_transformation[0].name
+  namespace           = aws_cloudwatch_log_metric_filter.sns-sms-phone-carrier-unavailable.metric_transformation[0].namespace
+  period              = 60 * 60 * 6
+  statistic           = "Sum"
+  threshold           = 10
+  alarm_actions       = [aws_sns_topic.notification-canada-ca-alert-warning.arn]
+  treat_missing_data  = "notBreaching"
+}
+
+resource "aws_cloudwatch_metric_alarm" "us-west-2-sns-sms-phone-carrier-unavailable-warning" {
```
Comment (QuanNguyen-CDS) [inline aws/common/cloudwatch_alarms.tf:191]: Same comment as above regarding defacto naming convention

**Review by QuanNguyen-CDS — COMMENTED** [review]
Overall the PR is very good.
One question: do we also want to monitor any of the other errors?
I don't know when this happens, "Phone carrier has blocked this message" but it might be another interesting piece of data

File: aws/common/cloudwatch_alarms.tf (hunk 2 above)
Comment (AntoineAugusti) [reply aws/common/cloudwatch_alarms.tf]: Haven't been very focused, sorry about that. Changed. Thanks for catching.

**Comment by AntoineAugusti** [comment]
> One question: do we also want to monitor any of the other errors?

Never saw "Phone carrier has blocked this message" yet but we may want to monitor that. Maybe later?

File: aws/common/cloudwatch_log.tf [hunk 4]
```diff
@@ -46,6 +49,72 @@ resource "aws_cloudwatch_log_group" "sns_deliveries_failures_us_west_2" {
...
+
+  metric_transformation {
+    name          = "sns-sms-blocked-as-spam"
+    namespace     = "LogMetrics"
+    value         = "1"
+    default_value = "0"
+  }
+}
+
+resource "aws_cloudwatch_log_metric_filter" "us-west-2-sns-sms-blocked-as-spam" {
```
Comment (QuanNguyen-CDS) [inline aws/common/cloudwatch_log.tf:69]: sorry what about this one?

File: aws/common/cloudwatch_log.tf [hunk 5]
```diff
@@ -46,6 +49,72 @@ resource "aws_cloudwatch_log_group" "sns_deliveries_failures_us_west_2" {
...
+
+  metric_transformation {
+    name          = "sns-sms-phone-carrier-unavailable"
+    namespace     = "LogMetrics"
+    value         = "1"
+    default_value = "0"
+  }
+}
+
+resource "aws_cloudwatch_log_metric_filter" "us-west-2-sns-sms-phone-carrier-unavailable" {
```
Comment (QuanNguyen-CDS) [inline aws/common/cloudwatch_log.tf:99]: same

**Review by QuanNguyen-CDS — APPROVED** [review]

**Pull request merged**